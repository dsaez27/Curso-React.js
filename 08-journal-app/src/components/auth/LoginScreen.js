import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { useForm } from './../../hooks/useForm';
import {
    startGoogleLogin,
    startLoginEmailPassword,
} from '../../redux/actions/auth';

export const LoginScreen = () => {
    const dispatch = useDispatch();
    const { loading } = useSelector((state) => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: 'Daniel',
        password: '123456',
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        isFormValid && dispatch(startLoginEmailPassword(email, password));
    };

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    };

    const isFormValid = () => {
        if (email.trim().length === 0) {
            return false;
        } else if (password.trim().length === 0) {
            return false;
        }

        return true;
    };

    return (
        <>
            <h3 className='auth__title'>Login</h3>

            <form onSubmit={handleLogin}>
                <input
                    className='auth__input'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />
                <input
                    className='auth__input'
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />
                <button
                    className='btn btn-primary btn-block'
                    type='submit'
                    disabled={loading}
                >
                    Login
                </button>
                <div className='auth__social-networks'>
                    <p>Login with social networks</p>
                    <div className='google-btn' onClick={handleGoogleLogin}>
                        <div className='google-icon-wrapper'>
                            <img
                                className='google-icon'
                                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                                alt='google button'
                            />
                        </div>
                        <p className='btn-text'>
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to='auth/register' className='link'>
                    Create new account
                </Link>
            </form>
        </>
    );
};
