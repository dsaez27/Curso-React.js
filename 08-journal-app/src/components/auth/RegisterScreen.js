import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
    return (
        <>
            <h3 className='auth__title'>Register</h3>

            <form>
                <input
                    className='auth__input mb-1'
                    type='text'
                    placeholder='Name'
                    name='name'
                    autoComplete='off'
                />
                <input
                    className='auth__input mb-1'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                />
                <input
                    className='auth__input mb-1'
                    type='password'
                    placeholder='Password'
                    name='password'
                />
                <input
                    className='auth__input mb-1'
                    type='password'
                    placeholder='Confirm Password'
                    name='password2'
                />
                <button
                    className='btn btn-primary btn-block mb-5'
                    type='submit'
                >
                    Register
                </button>

                <Link to='auth/login' className='link'>
                    Already registered?
                </Link>
            </form>
        </>
    );
};
