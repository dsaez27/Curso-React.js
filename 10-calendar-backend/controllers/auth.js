const User = require('../models/User');
const bcrypt = require('bcryptjs');
const { generateJwt } = require('../helpers/jwt');

const crearUsuario = async (req, res) => {
    const { email, password } = req.body;

    try {
        let user = await User.findOne({ email });
        console.log(user);
        if (user) {
            return res.status(400).json({
                ok: false,
                msg: 'User already exists',
            });
        }

        user = new User(req.body);

        //Encriptar contraseÃ±a
        const salt = await bcrypt.genSalt(10);
        user.password = await bcrypt.hash(password, salt);

        await user.save();

        //Generate token
        const token = await generateJwt(user._id, user.name);

        res.status(201).json({
            ok: true,
            uid: user._id,
            name: user.name,
            token,
        });
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Plase report this error to the admin',
        });
    }

    res.status(201).json({
        ok: true,
        msg: 'User created',
    });
};

const loginUsuario = async (req, res) => {
    const { email, password } = req.body;

    try {
        let user = await User.findOne({ email });
        console.log(user);
        if (!user) {
            return res.status(400).json({
                ok: false,
                msg: 'User not exists with this email',
            });
        }

        //Validate password
        const validPassword = await bcrypt.compare(password, user.password);

        if (!validPassword) {
            return res.status(400).json({
                ok: false,
                msg: 'Invalid password',
            });
        }

        //Generate token
        const token = await generateJwt(user._id, user.name);

        res.status(200).json({
            ok: true,
            uid: user._id,
            name: user.name,
            token,
        });
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Plase report this error to the admin',
        });
    }
};

const revalidarToken = async (req, res) => {
    const uid = req.uid;
    const name = req.name;

     //Generate token
     const token = await generateJwt(uid, name);

    res.json({
        ok: true,
        token
    });
};

module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken,
};
