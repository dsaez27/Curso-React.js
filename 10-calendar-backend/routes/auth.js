const { Router } = require('express');
const {
    crearUsuario,
    loginUsuario,
    revalidarToken,
} = require('../controllers/auth');
const router = Router();
const { check } = require('express-validator');
const { validarCampos } = require('../middlewares/validarCampos');
const { validarJwt } = require('../middlewares/validarJwt');

router.post(
    '/new',
    [
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'Password to be at least 6 characters').isLength({
            min: 6,
        }),
        validarCampos,
    ],
    crearUsuario
);
router.post(
    '/',
    [
        check('email', 'Email is required').isEmail(),
        check('password', 'Password to be at least 6 characters')
            .not()
            .isEmpty(),
        validarCampos,
    ],
    loginUsuario
);
router.get('/renew', validarJwt, revalidarToken);

module.exports = router;
