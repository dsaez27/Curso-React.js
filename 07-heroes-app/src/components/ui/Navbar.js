import React, { useContext } from 'react';
import { Link, NavLink, useHistory } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const Navbar = () => {

    const history = useHistory();

    const {
        user: { name },
    } = useContext(AuthContext);

    const { dispatch } = useContext(AuthContext);

    const handleLogout = () => {
        dispatch({ type: types.logout });
        localStorage.removeItem('user');
        history.replace('/login');
    };

    return (
        <nav className='navbar navbar-expand-lg navbar-dark bg-dark'>
            <div className='container-fluid'>
                <Link className='navbar-brand' to='/'>
                    <h3> Asociaciones </h3>
                </Link>
                <div className='collapse navbar-collapse' id='navbarNav'>
                    <div className='navbar-nav'>
                        <NavLink
                            activeClassName='active'
                            className='nav-item nav-link'
                            exact
                            to='/marvel'
                        >
                            Marvel
                        </NavLink>

                        <NavLink
                            activeClassName='active'
                            className='nav-item nav-link'
                            exact
                            to='/dc'
                        >
                            DC
                        </NavLink>
                        <NavLink
                            activeClassName='active'
                            className='nav-item nav-link'
                            exact
                            to='/search'
                        >
                            Search
                        </NavLink>
                    </div>
                </div>
                <div className='d-flex'>
                    <div className='navbar-nav me-2'>
                        <span className='nav-item nav-link text-info'>
                            {name}
                        </span>
                        <button
                            className='nav-item nav-link btn'
                            onClick={handleLogout}
                        >
                            Logout
                        </button>
                    </div>
                </div>
            </div>
        </nav>
    );
};
