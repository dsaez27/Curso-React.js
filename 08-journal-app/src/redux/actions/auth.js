import Swal from 'sweetalert2';
import { types } from '../types';
import { finishLoading, setError, startLoading } from './ui';
import { noteLogout } from './notes';
import { firebase, googleAuthProvider } from '../../firebase/firebaseConfig';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        try {
            return firebase
                .auth()
                .signInWithEmailAndPassword(email, password)
                .then(({ user }) => {
                    dispatch(login(user.uid, user.displayName));
                    dispatch(finishLoading());
                })
                .catch((e) => {
                    dispatch(finishLoading());
                    Swal.fire({
                        title: 'Error',
                        text: e.message,
                        icon: 'error',
                    });
                });
        } catch (e) {
            console.log(e);
        }
    };
};

export const startRegisterEmailPasswordName = (email, password, name) => {
    return async (dispatch) => {
        try {
            await firebase
                .auth()
                .createUserWithEmailAndPassword(email, password)
                .then(async ({ user }) => {
                    await user.updateProfile({
                        displayName: name,
                    });

                    dispatch(login(user.uid, user.displayName));
                })
                .catch(({ message }) => {
                    Swal.fire({ title: 'Error', text: message, icon: 'error' });
                    dispatch(setError(message));
                });
        } catch (e) {
            console.log(e);
        }
    };
};

export const startGoogleLogin = () => {
    return async (dispatch) => {
       
            await firebase
                .auth()
                .signInWithPopup(googleAuthProvider)
                .then(({ user }) => {
                    dispatch(login(user.uid, user.displayName));
                })
                .catch((error) => {
                    console.log(error);
                });
         
    };
};

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName,
        },
    };
};

export const startLogout = () => {
    return (dispatch) => {
        firebase.auth().signOut();

        dispatch(logout());
        dispatch(noteLogout());
    };
};

export const logout = () => ({
    type: types.logout,
});
