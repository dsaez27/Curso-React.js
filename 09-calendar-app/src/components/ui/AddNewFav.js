import React from 'react';
import { useDispatch } from 'react-redux';
import { uiOpenModal } from './../../actions/ui';

export const AddNewFav = () => {
    const dispatch = useDispatch();

    const handleClickNew = () => {
        dispatch(uiOpenModal());
    };

    return (
        <>
            <button className='btn btn-primary fab' onClick={handleClickNew}>
                <i className='fa fa-plus'></i>
            </button>
        </>
    );
};
