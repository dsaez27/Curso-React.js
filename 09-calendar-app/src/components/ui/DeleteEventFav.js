import React from 'react';
import { useDispatch } from 'react-redux';
import { eventDeleted } from './../../actions/events';

export const DeleteEventFav = ({ id }) => {
    const dispatch = useDispatch();
    const handleDeleteEvent = () => {
        dispatch(eventDeleted(id));
    };

    return (
        <button
            className='btn btn-danger fab-danger'
            onClick={handleDeleteEvent}
        >
            <i className='fas fa-trash-alt'></i>
            <span> Borrar Evento</span>
        </button>
    );
};
