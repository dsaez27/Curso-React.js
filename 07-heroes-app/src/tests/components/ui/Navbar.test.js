import { Navbar } from './../../../components/ui/Navbar';
import { AuthContext } from './../../../auth/AuthContext';
import '@testing-library/jest-dom';
import { MemoryRouter, Router } from 'react-router-dom';
import { mount } from 'enzyme';
import { types } from './../../../types/types';

describe('Pruebas en el <navbar />', () => {
    const historyMock = {
        push: jest.fn(),
        replace: jest.fn(),
        location: {},
        listen: jest.fn(),
        createHref: jest.fn(),
    };

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: true,
            name: 'Daniel',
        },
    };

    const wrapper = mount(
        <AuthContext.Provider value={contextValue}>
            <MemoryRouter>
                <Router history={historyMock}>
                    <Navbar />
                </Router>
            </MemoryRouter>
        </AuthContext.Provider>
    );

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.text-info').text().trim()).toBe('Daniel');
    });

    test('Debe dede llamar el logout y usar el history', () => {
        wrapper.find('button').simulate('click');
        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.logout,
        });
        expect(historyMock.replace).toHaveBeenCalledWith('/login');
    });
});
