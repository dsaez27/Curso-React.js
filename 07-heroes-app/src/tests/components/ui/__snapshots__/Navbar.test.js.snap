// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <navbar /> Debe de mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      onClick={[Function]}
    >
      <h3>
         Asociaciones 
      </h3>
    </a>
    <div
      className="collapse navbar-collapse"
      id="navbarNav"
    >
      <div
        className="navbar-nav"
      >
        <a
          aria-current={null}
          className="nav-item nav-link"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
    <div
      className="d-flex"
    >
      <div
        className="navbar-nav me-2"
      >
        <span
          className="nav-item nav-link text-info"
        >
          Daniel
        </span>
        <button
          className="nav-item nav-link btn"
          onClick={[Function]}
        >
          Logout
        </button>
      </div>
    </div>
  </div>
</nav>
`;
